### PR TITLE
migrating community vm info modules into guest_info

### PR DESCRIPTION
**Pull request by mikemorency**
##### SUMMARY
This is an attempt to migrate vm info modules from community.vmware (vmware_guest_info, vmware_vm_info) into vmware.vmware. Since there is already a vm info module in vmware.vmware (guest_info), the functionality from the two community modules has been added to guest_info

This module also requires the rest and soap APIs. So for the simulator tests i added an API gateway so both APIs can be accessed on the same port.

All of the changes should be backwards compatible. We do plan on deprecating the module's old output and removing it at some point in the future. More on that in the additional info section.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
guest_info

##### ADDITIONAL INFORMATION
While the inputs for guest_info has not changed, the output should probably be changed. Theres some duplicated information, and some of the values previously returned by `guest_info` have similar names. 

Here is how the guest_info module currently returns info:
```
guests: [
  {
    old values here
  }
]
```

For now, we are returning the old values in two locations to the user:
```
guests: [
  {
    old and new values mixed together
    .....
    identity: {
        old values here
    }
  }
]
```

At some point in the future, the old values _not_ nested under `identity` will be removed. Resulting in:
It makes more sense to me if the old values are nested in an attribute like:
```
guests: [
  {
    new values here
    ......
    identity: {
        old values here
   }
  }
]
```

### CODE REVIEW

**Comment by mariolenz** [comment]
@mikemorency Sorry, didn't find the time to review yet. Will try to do it tomorrow!

**Review by mariolenz — COMMENTED** [review]
I wanted to check if this could really be a replacement for `comunity.vmware.vmware_guest_info` and at first glance it looks OK. I'd like to have a closer look, though, and also would want to compare it with `comunity.vmware.vmware_vm_info`.

However, there's one thing I want to mention: `comunity.vmware.vmware_guest_info` works, but `vmware.vmware.guest_info` failed with:

> Failed to connect to vCenter or ESXi API at ****:443 due to SSL verification failure : HTTPSConnectionPool(host='****', port=443): Max retries exceeded with url: /api (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)')))

It worked with `validate_certs: false` but this is something that might to be addressed. I'll test again and will provide more information, at least about the certificate of the vCenter.

**Comment by mariolenz** [comment]
> But it's weird that I run into this SSL error. The certificate of the vCenter is signed by a private CA, but the OS trusts this CA. Both `curl` and `comunity.vmware.vmware_guest_info` work fine, but `vmware.vmware.guest_info` doesn't.

@mikemorency Looks like something connected to [requests](https://github.com/psf/requests) / [certifi](https://github.com/certifi/python-certifi). As far as I can see, it looks like they think "it's not a bug but a feature" to use their own set of trusted CAs. ATM I tend to disagree...

File: .github/workflows/ansible-test.yml [hunk 1]
```diff
@@ -18,7 +18,10 @@ jobs:
         run: |
           set -euxo pipefail
           sudo apt-get update
-          sudo apt-get install podman
+
+          # recent version of podman has a bug. Fix is waiting deployment
```
Comment (bardielle) [inline .github/workflows/ansible-test.yml:22]: Is there a ticket in our kanban tracking it? 
If no, please open

File: plugins/module_utils/vmware.py [hunk 2]
```diff
@@ -403,3 +404,22 @@ def get_resource_pool_by_name(self, pool_name, fail_on_missing=False):
         if not pool and fail_on_missing:
             self.module.fail_json("Unable to find resource pool with name %s" % pool_name)
         return pool
+
+    def get_all_objs_by_type(self, vimtype, folder=None, recurse=True):
```
Comment (bardielle) [inline plugins/module_utils/vmware.py:408]:
```suggestion
    def list_objs_by_type(self, vimtype, folder=None, recurse=True):
```

Comment (bardielle) [inline plugins/module_utils/vmware.py:408]: please add description

File: plugins/module_utils/vmware_facts.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        'hw_is_template': vm.config.template,
+        'hw_folder': None,
+        'hw_version': vm.config.version,
+        'instance_uuid': vm.config.instanceUuid,
+        'guest_tools_status': _get_vm_prop(vm, ('guest', 'toolsRunningStatus')),
+        'guest_tools_version': _get_vm_prop(vm, ('guest', 'toolsVersion')),
+        'guest_question': json.loads(json.dumps(vm.summary.runtime.question, cls=VmomiSupport.VmomiJSONEncoder,
+                                                sort_keys=True, strip_dynamic=True)),
+        'guest_consolidation_needed': vm.summary.runtime.consolidationNeeded,
+        'ipv4': None,
```
Comment (bardielle) [inline plugins/module_utils/vmware_facts.py:62]: Why does this value is None? (also ipv6)

File: plugins/module_utils/vmware_facts.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+    for attribute in attributes:
+        try:
+            result = getattr(result, attribute)
+        except (AttributeError, IndexError):
+            return None
+    return result
+
+
+def gather_vm_facts(content, vm):
+    """ Gather facts from vim.VirtualMachine object. """
```
Comment (bardielle) [inline plugins/module_utils/vmware_facts.py:37]: This is long fucntion, please refactoring it into several function

**Comment by machacekondra** [comment]
> > But it's weird that I run into this SSL error. The certificate of the vCenter is signed by a private CA, but the OS trusts this CA. Both `curl` and `comunity.vmware.vmware_guest_info` work fine, but `vmware.vmware.guest_info` doesn't.
> 
> @mikemorency Looks like something connected to [requests](https://github.com/psf/requests) / [certifi](https://github.com/certifi/python-certifi). As far as I can see, it looks like they think "it's not a bug but a feature" to use their own set of trusted CAs. ATM I tend to disagree...

I think it's worth to at least mention in documentation how to use system wide certificates, if don't support it by some parameter, but most of the users would be expect it to be used IMHO, they are not aware that we use `requests` library, so we should mention it in doc.

**Review by machacekondra — APPROVED** [review]

**Comment by mikemorency** [comment]
The REST util is the only thing that relies on requests right now. I think that is why mario experienced the issue with this module and not the community one, it uses pyvmomi.

I added a note to the rest api docs. The user can set an environment variable to specify the ca bundle to use (instead of the certifi one provided).
REQUESTS_CA_BUNDLE, https://requests.readthedocs.io/en/latest/user/advanced/

**Comment by mariolenz** [comment]
> I added a note to the rest api docs. The user can set an environment variable to specify the ca bundle to use (instead of the certifi one provided). REQUESTS_CA_BUNDLE, https://requests.readthedocs.io/en/latest/user/advanced/

Excellent! Thanks!

**Review by mariolenz — APPROVED** [review]
LGTM

**Review by bardielle — APPROVED** [review]

**Pull request merged**